src/query/mlt/mlt.rs

```diff
@@ -0,0 +1,387 @@
+    fn is_noise_word(&self, word: String) -> bool {
+        let word_length = word.len();
+        if word_length == 0 {
+            return true;
+        }
+        if self
+            .min_word_length
+            .map(|min| word_length < min)
+            .unwrap_or(false)
+        {
+            return true;
+        }
+        if self
+            .max_word_length
+            .map(|max| word_length > max)
+            .unwrap_or(false)
+        {
+            return true;
+        }
+        return self.stop_words.contains(&word);
+    }
+
+    /// Couputes the score for each term while ignoring not useful terms
+    fn create_score_term(
+        &self,
+        searcher: &Searcher,
+        per_field_term_frequencies: HashMap<Term, usize>,
+    ) -> Result<Vec<ScoreTerm>> {
+        let mut score_terms = BinaryHeap::new();
+        let num_docs = searcher
+            .segment_readers()
+            .iter()
+            .map(|x| x.num_docs() as u64)
+            .sum::<u64>();
+
+        for (term, term_frequency) in per_field_term_frequencies.into_iter() {
+            // ignore terms with less than min_term_frequency
+            if self
+                .min_term_frequency
+                .map(|x| term_frequency < x)
+                .unwrap_or(false)
+            {
+                continue;
+            }
+
+            let doc_freq = searcher.doc_freq(&term)?;
+
+            // ignore terms with less than min_doc_frequency
+            if self
+                .min_doc_frequency
+                .map(|x| doc_freq < x)
+                .unwrap_or(false)
+            {
+                continue;
+            }
+
+            // ignore terms with more than max_doc_frequency
+            if self
+                .max_doc_frequency
+                .map(|x| doc_freq > x)
+                .unwrap_or(false)
+            {
+                continue;
+            }
+
+            // ignore terms with zero frequency
+            if doc_freq == 0 {
+                continue;
+            }
+
+            // compute similarity & score
+            let idf = self.idf(doc_freq, num_docs);
+            let score = (term_frequency as f32) * idf;
+            score_terms.push(ScoreTerm::new(term, score));
+        }
+
+        // limit ourself to max_query terms. we need to sort so to avoid discarding important terms
+        let score_terms = if let Some(max_query_terms) = self.max_query_terms {
+            let max_num_terms = std::cmp::min(max_query_terms, score_terms.len());
+            score_terms
+                .into_sorted_vec()
+                .into_iter()
+                .take(max_num_terms)
+                .collect()
+        } else {
+            score_terms.into_sorted_vec()
+        };
+        Ok(score_terms)
+    }
+
+    /// Computes the similarity
+    fn idf(&self, doc_freq: u64, doc_count: u64) -> f32 {
+        let x = ((doc_count - doc_freq) as f32 + 0.5) / (doc_freq as f32 + 0.5);
```

Comment:
no sure if it is possible, but have  you check if we could reuse the logic in the BM25 object?

Reply:
I will check

Reply:
(it might not be possible.. i did not investigate)